Merging help

turican0 opened the pull request:
This is a package with many modifications and fixes, mainly to the renderers.
I changed some of the little things, if I thought there was a better way to do it differently, but there would be a good reason to do it as it was originally, I'm not opposed to changing it.

First, I integrated the changes made by grimsqueaker.

I also integrated all of your edits into it, if possible:

Render refactor #93
-------------------
feat: implemented new interface class and openGL graphics setting. - integrated
fix: fixed error in type checking code - integrated
Merge branch 'development' into opengl-game-render - nothing changes
Merge branch 'development' into opengl-game-render - nothing changes
Merge branch 'development' into opengl-game-render - nothing changes
Merge branch 'development' into opengl-game-render - nothing changes
fix: fixed out dated method names - not integrated - I renamed the function names with the correct addresses
refactor: Overloaded VGA_Blit to use m_iOrig but also allow use of pa… - not integrated - There is no need to send the height and width parameters with each Blit, just send them when the resolution changes.
refactor: implemented interface for GameRender's - integrated
Merge branch 'development' into render-refactor - nothing changes
feat: added skeleton of OpenGL render class - integrated
feat: added new openGLRender flag and - integrated (GL class renamed to GameRender_GL)
fix: added GLGameRender class to cmake list - integrated

Other changes:
Created three skeletons:
GameRender_old
GameRender_new
GameRender_GL
(I put the renderer type last, for easier alphabetical searching)

The viewPort setting is not renderer dependent, so I put it on the outside. Now only the DrawWorld function is called in the renderer.

I left the function calls as in the original game, which will make it easier to find bugs. I moved the other variables mostly to Basic.cpp, as global variables. This allows toggling the render on the fly and therefore easier bug finding. Also, I don't know if it was ideal to have so many variables in the function calls. If you want to keep track of what variables are used in the functions, wouldn't it be enough to put them in a comment before the function?

At most I tried to unify the code of the old and new renderer, which also led to a few minor fixes on both sides. Apart from a few minor things, the main difference now is that the new one can do arbitrary resolutions and multiple cores

I put the initialization and deinitialization of the renderer right before and after the sub_47320_in_game_loop function, which I find more readable and less error-prone.

I've added an ifdef TEST_RENDERERS that can be used to compare the result of both renderers.

Bug fixes(some may have been introduced during unification):
-fixes viewport behavior in some cases
-map not displaying large(when pressing enter) - missing SetRenderViewPortSize_40BF0 before DrawWorld_411A0
-error at the end of level 5 when going to secret level in DrawInverseSquareInProjectionSpace missing line x_DWORD_DE55C_ActTexture = x_DWORD_DDF50_texture_addresses. at(Str_E9C38_smalltit[index].byte41); if DrawInverseSquareInProjectionSpace was started before DrawSquareInProjectionSpace was initialized, the variable x_DWORD_DE55C_ActTexture

Bugs I encountered:
-at atypical resolution, when the spell icons in the UI are rendered separately at the top right, transparency and hiding don't work on them
-the game uses one more process than in the settings (X+1)
-I'm not sure if the viewport settings (384, 0, 256, 400) will work for arbitrary resolutions, I guess we'll have to recalculate from the resolution

Now I'll take a moment to comment, approve the changes. I have to edit all the changes in my developing branch and then I'll start unifying sub_main.cpp, I'll try to avoid renderers because you may have to change something for GL renderer or fixes.

Please let me know what you guys think of the changes.

Have a great holiday season and good luck with your future development.

Tom.

turican0 commented:
GameRender.cpp replaced by GameRender_old.cpp, GameRender_new.cpp and GameRender_GL.cpp

GrimSqueaker reviewed (dismissed):
PR looks fine. Playing through level 1 in 32 and 64 bit on Linux works.

turican0 commented:
That's great. It's possible that in 64-bit I'll still run into other things in sub_main.cpp that only show up in certain parts of the game, or in the binary compatibility of saves, etc.
Anyway, I'll keep on merging.
Tom.

turican0 commented:
Everything is merged !!! Casually tried the game. It works, but in some resolutions there is a problem with mouse position and AI rendering.
But it's nothing that shouldn't be easy to fix.
I'll address this as a priority(but I'm on vacation from Wednesday to Friday, so I'll fix something then)

Because of the network, I had to add the NetworkLib library in NuGet I couldn't find it. If the Actions on GitHub won't work, or you find a way to compile the game better, please fix it. Thank you very much.

I'm glad we have one repository now.
After I fix the basic bugs, I'll play around with the network, and then run the game through the regression tests again.

Hi, Tom.

Translated with www.DeepL.com/Translator (free version)

GrimSqueaker commented:
Great. Thank you. 
On the first look there seem to be a few regressions, but nothing we shouldn't be able to fix.

turican0 commented:
First bug(move in 320x200) fixed!

turican0 commented:
I discovered debug on screen informations in the game code. I add activation using the "J" key. Most of the stuff is displayed incorrectly, but I'll try to bring up the menu in the DOS version and fix what I can.

thobbsinteractive commented:
Hi guys. So I will look at this, I was planning to on Sunday, but my flight back was delayed. Now it looks like I have COVID. The symptoms are not bad, but I am a bit tired. 
I am surprised about the Mouse/button alignments. They are all working correctly in a range of resolutions and 320x200. Hopefully the Sunday commits fix this. I will give it a good long test tomorrow

turican0 commented on remc2/utilities/BitmapIO.h:
```diff
@@ -20,7 +20,7 @@ class BitmapIO
 	static void WriteImageBufferAsImageBMP(char* path, int width, int height, uint8_t* ptrPalette , uint8_t* ptrBuffer);
 	static void WriteRGBAImageBufferAsImageBMP(char* path, int width, int height, uint8_t* ptrPalette, uint8_t* ptrBuffer);
-	static void WritePaletteAsImageBMP(char* path, int numColors, uint8_t* ptrPalette);
+	static void WritePalletteAsImageBMP(char* path, int numColors, uint8_t* ptrPalette);
```
It clearly has one "l". I got confused and thought someone else put the double "l" in there. Anyway, now the double l is everywhere and it's best to globally rename it to single l.

turican0 replied:
Now I'll wait to see what you have to say about my modifications. In the meantime, I'll play with the network.
I've got some minor ROL/ROR optimization ready for the renderers. Subjectively it will add 1-2FPS.
inline uint32 __ROL4_16__(uint32 value) { return (value << 16) | (value >> 16); }
But I'll put that one in with other changes. We could also get rid of __ROL4_16__ altogether, it's about swapping WORDS in DWORD.

GrimSqueaker replied:
Welcome back. I hope that you have a mild infection and get well soon.

I stumbled across the ROL before, because the sanitzer does not like it when the shifts loose bits. It does not realize that in this case it is intentional. It would be better to mask with `0xffff0000` and `0x0000ffff`.

In preliminary tests I also had problems with cursor position. But this might be due to running on a 4k notebook display on 200% scaling which essentially scales the SDL scaled image again. I need to look into this again.

thobbsinteractive commented:
I cannot seem to compile it in Visual Studio. It keeps telling me that` '_vsnprintf': is not a member of 'std'`
This is strange because it was introduced in ISO C++ 11 and we are compiling to ISO C++ 14.
Also I cannot push to this branch @turican0 can you give me permission on this branch?

thobbsinteractive commented:
Oh yes. If you up scale more than 100% (in Windows) It breaks the cursor
alignment. It has been like that for along time. I did occure to me this
would be a problem for 4k where it is normal to increase zoom. But I never
logged it. My bad

On Mon, 11 Jul 2022, 21:59 GrimSqueaker, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In remc2/utilities/BitmapIO.h
> <https://github.com/thobbsinteractive/magic-carpet-2-hd/pull/98#discussion_r918301254>
> :
>
> > @@ -20,7 +20,7 @@ class BitmapIO
>
>  	static void WriteImageBufferAsImageBMP(char* path, int width, int height, uint8_t* ptrPalette , uint8_t* ptrBuffer);
>  	static void WriteRGBAImageBufferAsImageBMP(char* path, int width, int height, uint8_t* ptrPalette, uint8_t* ptrBuffer);
> -	static void WritePaletteAsImageBMP(char* path, int numColors, uint8_t* ptrPalette);
> +	static void WritePalletteAsImageBMP(char* path, int numColors, uint8_t* ptrPalette);
>
> Welcome back. I hope that you have a mild infection and get well soon.
>
> I stumbled across the ROL before, because the sanitzer does not like it
> when the shifts loose bits. It does not realize that in this case it is
> intentional. It would be better to mask with 0xffff0000 and 0x0000ffff.
>
> In preliminary tests I also had problems with cursor position. But this
> might be due to running on a 4k notebook display on 200% scaling which
> essentially scales the SDL scaled image again. I need to look into this
> again.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/thobbsinteractive/magic-carpet-2-hd/pull/98#discussion_r918301254>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AETQQFCAYKKIEVS35BJN4ATVTR4JLANCNFSM52ZUGNWQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

turican0 commented:
I've added you as a collaborator, but I don't know if that's enough. Please don't approve pull requests on my repository if there are code changes, it doesn't matter for compilation method changes.

Locally I rewrote std::snprintf to snprintf, and then it worked, but it only shows up in the local instance. Couldn't this be solved with a suitable location using namespace std;?

The fallback is a copy of boost in the project, but we want to avoid that.

GrimSqueaker:
Masking 0xffff0000 and 0x0000ffff, but you still have to add << a >> or temp=x.word[0];x.word[1]=x.word[0];x.word[1]=temp;, but (x << 16)|(a>>16) is probably the fastest.

Tom.

Translated with www.DeepL.com/Translator (free version)

turican0 commented:
Just a note my current branch is now "merging-help" (I'll rename it later). Ignore the "development" branch for now.

turican0 commented:
If you were fixing mouse behavior in scaled resolutions, it would be a good idea to put the code in port_vga_mouse.cpp, not in the game code.

thobbsinteractive commented:
Hey Tom, a "collaborator" is perfect. You might want to add @GrimSqueaker (if you have not already). Assuming you have the pull request rules setup to require 1 approval, then we will not be able to merge pull requests unless you request us as a reviewer. Since we are not admins we cannot force an approval either. So it is all good.

thobbsinteractive commented:
An assembly render would be faster, such things have been used to get Quake running on the Gameboy Advance. It would be cool to experiment with it. 
Obviously OpenGL give use a lot more speed. With all the filtering turned off it would also look much the same (except there would be a lot less texture distortion).
I see that NetworkLib is the repo from here? https://github.com/DarkWanderer/NetworkLib?
Since it does not seem to be in Nuget, maybe we can put pulling it in the build instructions and remove it from the Repo? Unless you have made changes to it @turican0 we are essentially duplicating code.

turican0 commented:
I'm not against loading from https://github.com/DarkWanderer/NetworkLib. I just don't know how to set it up. But for now, please wait until I'm further along with this, I want to make sure I don't have to change anything in the library so we don't redo it unnecessarily.

thobbsinteractive commented:
So the targeting is off. The fire balls fly right at their target, but only occasionally impact. 
The Viewport is not scaling on the [Enter/Map Screen]. 
Finally, the FPS seems much faster but we might need the frame limiter back as it we are (depending on what the player is doing) hitting between 25fps to 80 fps. Ideally we should lock it at around 30fps (bullfrog probably planned for around 25fps)

thobbsinteractive commented:
Also the pause menu text does not show in full HD, but this might be an existing scaling issue.

thobbsinteractive commented on remc2/engine/Basic.cpp:
```diff
@@ -2026,9 +2042,13 @@ void DrawLineHighRes(int16_t posStartX, int16_t posStartY, int16_t posEndX, int1
 	v6 = posEndY;
-	v10 = (unsigned __int16)(pitch - posEndX);
+	v10 = (unsigned __int16)(helpWidth - posEndX);
 	if (x_WORD_E36D4 & 4)
```
I am not sure about the name "helpWidth". Pitch is graphics term for basically line resolution in the buffer. We could use method overloading here? We can check DefaultResolutions in the overload method before calling the original method with a fix pitch of 640?

thobbsinteractive commented on remc2/engine/GameRenderInterface.h:
```diff
@@ -0,0 +1,36 @@
+	//virtual GameRenderInterface(uint8_t renderThreads, bool assignToSpecificCores, int rendererType);
+	//GameRenderInterface() {}
+	virtual ~GameRenderInterface() {}
+	//virtual void Init(uint8_t renderThreads, bool assignToSpecificCores) = 0;
```
We have a complete history, I think (where we can) we should delete commented out lines

turican0 commented:
1) On targeting errors. ViewPort in Enter. Mouse sensitivity. And text in Pause menu I'll look at on the weekend(I'm going away tomorrow).
2)Yes the FPS limiter could be at 30FPS. The code is already on the ends of VGA_BlitAny/sub_75200_VGA_Blit640/sub_90478_VGA_Blit320.
3)I'd like to avoid overloading - ideally, where it won't cause problems I'd keep the original call. The game was programmed by far more experienced programmers than us. They could have made mistakes too, but it's better for debugging. I guess it doesn't matter if it's pitch or width. It's originally width, but it's still used as pitch.
The helpWidth is what I meant to convey, it's an auxiliary variable that changes depending on the resolution used.
	int helpWidth = 640;
	int helpHeight;
	if (x_WORD_180660_VGA_type_resolution == 1)
		helpHeight = 400;
	else
		helpHeight = 480;
	if (!DefaultResolutions())
	{
		helpWidth = screenWidth_18062C;
		helpHeight = screenHeight_180624;
	}
4)Yes we can delete the commented rows. Just please leave the lines of type:
	//address 1f3783
#ifdef DEBUG_SEQUENCES
		//add_compare(0x1F3783, debugafterload);
#endif //DEBUG_SEQUENCES
	/*if (debugcount_sub_12780 >= 2)
	{
		debugcount_sub_12780++;
		debugcount_sub_12780--;
	}
	debugcount_sub_12780++;*/
These are handy for any debugging.
5) Ok, I prefer GameRenderGL, GameRenderHD and GameRenderOriginal.

thobbsinteractive commented:
> 1. On targeting errors. ViewPort in Enter. Mouse sensitivity. And text in Pause menu I'll look at on the weekend(I'm going away tomorrow).

Thanks. I would appreciate it very much!

>    2)Yes the FPS limiter could be at 30FPS. The code is already on the ends of VGA_BlitAny/sub_75200_VGA_Blit640/sub_90478_VGA_Blit320.

I can actually look into this. I remember how you did it before

>    3)I'd like to avoid overloading - ideally, where it won't cause problems I'd keep the original call. The game was programmed by far more experienced programmers than us. They could have made mistakes too, but it's better for debugging. 

Well yeah, they were the best of the best, but it is reversed engineered code right? So given the amount of duplication in the render, my guess is that it did not look anything like what we are working with. I wish we could get the source for EA.

> I guess it doesn't matter if it's pitch or width. It's originally width, but it's still used as pitch.
>    The helpWidth is what I meant to convey, it's an auxiliary variable that changes depending on the resolution used.
>    int helpWidth = 640;
>    int helpHeight;
>    if (x_WORD_180660_VGA_type_resolution == 1)
>    helpHeight = 400;
>    else
>    helpHeight = 480;
>    if (!DefaultResolutions())
>    {
>    helpWidth = screenWidth_18062C;
>    helpHeight = screenHeight_180624;
>    }
>    4)Yes we can delete the commented rows. Just please leave the lines of type:
>    //address 1f3783
>    #ifdef DEBUG_SEQUENCES
>    //add_compare(0x1F3783, debugafterload);
>    #endif //DEBUG_SEQUENCES
>    /_if (debugcount_sub_12780 >= 2)
>    {
>    debugcount_sub_12780++;
>    debugcount_sub_12780--;
>    }
>    debugcount_sub_12780++;_/
>    These are handy for any debugging.

Yeah, there is no way I would mess with your debug code. Or at least I would try my best not to

> 2. Ok, I prefer GameRenderGL, GameRenderHD and GameRenderOriginal.

Thanks at lot. Since you are away, I will make the change. We can always revert it if necessary.

turican0 commented:
errors:
-scaling works for me at 320x200 / 640x480, I'll try higher at the weekend
-ViewPort and Enter in 320x200 ok, 640x480 same as original in DOSBOX, can we give 480 height for 640x480,
but in the original it's 400, but over full screen height is better.
(line SetViewPortScreenCoordinates_2CA60(384, 0, 256, 400); but it would have to be changed somewhere else - I'll look into it over the weekend)
-mouse sensitivity so far - not tested
-text in Pause menu - yes, I'll check it this weekend

Generally I'll check it more at the weekend, but if I don't find anything for some bug I'll need more information.

GrimSqueaker commented on remc2/engine/GameRender_new.cpp:
```diff
@@ -1,28 +1,21 @@
 
-GameRender::GameRender(uint8_t* pScreenBuffer, uint8_t* pColorPalette, uint16_t screenWidth, uint16_t screenHeight, uint16_t viewPortPosX, uint16_t viewPortPosY, uint16_t viewPortWidth, uint16_t viewPortHeight, std::array<uint8_t*, 256> &textureAdresses, uint8_t pX_BYTE_F6EE0_tablesx[], uint8_t renderThreads, bool assignToSpecificCores)
+GameRender_new::GameRender_new(uint8_t renderThreads, bool assignToSpecificCores)
 {
```
> GrimSqueaker: Masking 0xffff0000 and 0x0000ffff, but you still have to add << a >> or temp=x.word[0];x.word[1]=x.word[0];x.word[1]=temp;, but (x << 16)|(a>>16) is probably the fastest.

I meant this: ((x & 0x0000ffff) << 16)|((x & 0xffff0000) >>16)

turican0 commented:
((x & 0x0000ffff) << 16)|((x & 0xffff0000) >>16)
and
(x << 16)|(x >>16)
both generated same asm code - 12 lines
        push    rbp
        mov     rbp, rsp
        mov     DWORD PTR [rbp-4], edi
        mov     eax, DWORD PTR [rbp-4]
        sal     eax, 16
        mov     edx, eax
        mov     eax, DWORD PTR [rbp-4]
        shr     eax, 16
        or      eax, edx
        pop     rbp
        ret

GrimSqueaker commented:
> ((x & 0x0000ffff) << 16)|((x & 0xffff0000) >>16) and (x << 16)|(x >>16) both generated same asm code - 12 lines push rbp mov rbp, rsp mov DWORD PTR [rbp-4], edi mov eax, DWORD PTR [rbp-4] sal eax, 16 mov edx, eax mov eax, DWORD PTR [rbp-4] shr eax, 16 or eax, edx pop rbp ret

Exactly what I would have expected. The difference is that the code with the masks makes the intention clear.

GrimSqueaker commented:
> Ok, I'm not going to get into it :) What I've tested so far, I also found it slower :) But my colleagues used the KeeperFX assembler: https://github.com/dkfans/keeperfx/blob/master/src/bflib_render_gpoly.c I'll try the code for fun.

Another problem with assembly is that it makes the code non-portable. Without assembly the code should also build just fine on ARM (Raspberry Pi, MacOS).

thobbsinteractive commented:
> > Ok, I'm not going to get into it :) What I've tested so far, I also found it slower :) But my colleagues used the KeeperFX assembler: https://github.com/dkfans/keeperfx/blob/master/src/bflib_render_gpoly.c I'll try the code for fun.
> 
> Another problem with assembly is that it makes the code non-portable. Without assembly the code should also build just fine on ARM (Raspberry Pi, Mac).

That's a really good point.  It is very much platform specific.

In other news, setting the gameRes in the ini file to 640x480 now instantiates and runs the GameRenderOriginal. I just test this but this render seems to have an issue drawing the sky.

turican0 commented:
Thanks for the info, I'll look into it this weekend as well, I may have made a mistake when consolidating. I haven't tried all the combinations yet.

thobbsinteractive commented:
> Thanks for the info, I'll look into it this weekend as well, I may have made a mistake when consolidating. I haven't tried all the combinations yet.

It cool man. If you see me posting more stuff just know it is so we have it recorded, it is not that I expect you to immediately work on it or reply. Have a good week. I hope this trip your going on is a fun one

thobbsinteractive commented:
Something is wrong with the terrain. These screen shots were taken from the GOG version and this branch on Level 5
![image](https://user-images.githubusercontent.com/19335188/178576552-bf0dbff8-6269-414a-b282-65ab89581fc6.png)

turican0 commented:
Hey, I'm back. I've been to the area where the great game "Someday You'll Return" is set.
(which, miraculously, is 90% made by two people)

turican0 commented:
I will start fixing some bugs, but it will be good to enter them as Issues so we can address them individually.

turican0 commented:
[Fix Linux build]

-see my note on refactoring: change ViewPort to a class

[fix: reimplemented SetRenderViewPortSize_BCD45 that takes a viewPort …]

It should not be if (Default Resolutions()), but if (!Default Resolutions()). I would see it more like this:
if (!DefaultResolutions())
code for Alt resolution
else
if (x_WORD_180660_VGA_type_resolution == 1)
code for 320x200
else
code for 640x480

refactor: change ViewPort to a Class. Put it in as a member of the Ga… 

-I don't think the ViewPort should be part of the renderer(ViewPort m_viewPort;). If we switch between renderers, we will unnecessarily have to add code at runtime to back up the viewport before cancelling the renderer and then passing it to the next renderer. In doing so, the ViewPort can be the same for all renderers. Additionally, we deviate from the original sub_5C1B0_set_any_variables2() procedure that sets the viewport by pre-creating a renderer. I would revert this commit.

chore: renamed definion checks to match new class names


-OK
Delete remc2.vcxproj.FileListAbsolute.txt 


-OK
Adjust CMakeLists.txt to file renamings


-OK
fix: added FileListAbsolute.txt to ignore list

-OK
fix: added FileListAbsolute.txt to ignore list

-OK
fix: renamed GameRender_old to GameRenderOriginal. GameRender_new to … 


-OK
fix: removed #define vsnprintf _vsnprintf from 
-OK
fixL renamed all instances of Pallette to Palette
-OK

thobbsinteractive commented:
Wait what? You just undid hours of my work, I fixed the map screen in that check-in. Can you talk to me before you revert my code please? Because I think we should put it back in and here is why:
We will not likely be changing the Game renderer mid game, to do so we would have to implement a new menu item or keypress. So I see absolutely no need for that. Games in general have not let you do that since the late 90s. So ViewPort can be a member of the Renders. That is obeying a principle of OOP encapsulation. ViewPort has a UML "Has a" relationship to all the renders. We also avoid more Global variables and that in turn makes the code easier to re-use, more flexible and even Unit Test (should we ever do Unit Tests to that level).
In regards to DefaultResolutions()
I was always taught where possible to have only have Positive logic in the first part of an If statement. So `If (DefaultResolutions())` is not only easier to read the flow of but it also puts your Original Render code first @turican0 in the code which I thought you would prefer.

@GrimSqueaker Where are you on this?

turican0 commented:
I understand that you want to encapsulate it, but the more we change the original c principles to c++ principles, the harder it will be to find bugs. ViewPort is more of a UI setting that has nothing to do with the renderer. It can be a class, but it shouldn't be part of the renderer, but some higher class. 

Otherwise you can't use the original code:
signed int sub_5C1B0()
...
  sub_87090();
  if ( byte_180552 )
    *(_BYTE *)(dword_D41A4 + 10) = 1;
  sub_2CA90(*(char *)(dword_D41A0 + 8589)); ---!!!!
  return 1;
}

I implemented the renderer switching mainly for testing purposes, but a lot of remakes also have a switch for classic/enhanced look. 

For the width settings, I'd like to keep 320x200 and 640x480 - the same as the original and an alternative for higher resolutions.
DefaultResolutions() is 320x200 and 640x480. !DefaultResolutions() is higher/alternate resolution.
Using the 'R' key would toggle either:
640 and 320.
Or
Alternate and 320.

I'm not defending the discussion, but a viewport in the renders would complicate some things. If necessary, I'd come back to it later when the game is in a more debugged state.

thobbsinteractive commented:
For now I will leave the code alone until you let me know when you are finished for that day. Otherwise we will have more potential merge issues.

turican0 commented:
Okay, I'm not planning too many changes right now. I mainly wanted to fix the bugs. Now I'm looking to fix RenderHD with a scaled viewport. For 320x200 and 640x480 resolutions and the old renderer it works, so either something is different in the renderer or the resolution values are hardwired somewhere.
In general if possible I would keep the logic. And if we want to add something, there should be extra features that don't change the original code. But it wants to consider if we need to make new functions or just fix/extend the original code.
I'll look at the blinking target later.
We can change the map size for alt resolution - or make it parametric.

thobbsinteractive commented:
> In the end, the bad rendering of the shrunken viewport was just due to an unrounded Y value of the viewport position :) I won't do anything today. You can change the ViewPort to a global class. But please try to change as little as possible. You can even try to implement a different map dimension. Tomorrow I'll look for more bugs, but as I wrote I'm losing track of it, better I'll start putting them in Issues. Thanks and don't be mad if I returned your code, otherwise I wouldn't have figured out why it behaves differently than it should.

I was looking into this yesterday and began to think it was a rounding issue too. I did have it working at one point I believe on my main branch using a lot of new code (because as you know the original using two fixed resolutions) and it was using rounding (if I remember correctly). I will make the changes to ViewPort tonight. Have a nice evening

thobbsinteractive commented:
![image](https://user-images.githubusercontent.com/19335188/179289748-3290b285-3790-49b1-a705-5d5c83bfd491.png)
This is likely caused by the GameRenderHD referencing a global variable that is maybe not changed when 'R' is press? That is my best guess

GrimSqueaker commented:
Hey guys,

> We will not likely be changing the Game renderer mid game,

I also think that this is not really necessary. The renderer can be selected nicely in the config file. Switching resolutions in-game was a nice and important feature in the 90s. But aside from debugging and comparing to the original rendering I see no point in keeping this. I think at some point we need to break 100% compatibility to the original in order to move forward with modernizations.

> We also avoid more Global variables and that in turn makes the code easier to re-use, more flexible and even Unit Test

Agree. In my first attempt of cleaning up the repo (https://github.com/GrimSqueaker/remc2clean) I started to replace all file handling with a single class that provides all the assets. This way it removes global variables and becomes testable. I would like to do the same thing here at some point. In fact I wanted to start with that already, but this big PR kind of blocks this for the moment.

> Using the 'R' key would toggle either:
640 and 320.
Or
Alternate and 320.

What is the point in the 320 resolution nowadays? I can understand that for nostalgic reasons one might want to play in up-scaled 640x480, but I guess that most people would prefer HD over any other setting and do not care about 320x200 at all.

In general I think we should get this PR merged as soon as possible so that all experiments (reverts, etc.) can take place on branches. This way an experimental revert can be done without breaking stuff. If there are just minor issues remaining we can also fix them with future PRs.

turican0 commented:
Ok, in the future we can keep only one resolution, as well as make larger adjustments to the variables. However, before that, please give me some time to fix the major issues (14 days ?) Once the major adjustments are made, I won't be able to fix some bugs because the data won't compare with the original game. The part of the code we are changing now is really only a fraction of the whole game, AND the question is how much work it would be to rewrite the whole game into code that we can understand and can fix the bugs without comparing it to the original game.

thobbsinteractive commented:
Hey Tom, 
Nice work on the mini-map, unfortunately the map entities are still drifting a bit.

![image](https://user-images.githubusercontent.com/19335188/179358575-052edf3e-66c1-4339-8c79-f871655f40e7.png)

thobbsinteractive commented:
Once we get this and the other stuff fix we can merge. My biggest concern is the Mini-map and the Terrain error (as they directly impact gameplay). 
We might want to refactor the mini-map into its own class one day, but not in this PR :)

turican0 commented:
I'm still trying to get branching (x_WORD_180660_VGA_type_resolution == 1) and DefaultResolutions() to work correctly there are still combinations where it doesn't work, which affects switching between 320 and alternative resolutions. Then I look at Minimap and finally at the terrain issue.

turican0 commented:
I returned the height of the minimap to 400 pixels. The problem is not with the entities, but with the map layer. The minimap rotation is hardcoded for a specific aspect ratio, the rendering code would have to be rewritten for other ratios.

thobbsinteractive commented:
> I returned the height of the minimap to 400 pixels. The problem is not with the entities, but with the map layer. The minimap rotation is hardcoded for a specific aspect ratio, the rendering code would have to be rewritten for other ratios.

That is cool. That makes sense. If we refactor it one day we can take a look at drawing it at different scales. It works now.

thobbsinteractive commented:
Yep 320x200 for Alt resolution seems to be working again. Map is fine, menus all work. Nice work buddy

thobbsinteractive commented:
Okay so it is looking really stable now, but the Cave Levels are using the same Texture for the floor and Ceiling. As I remember there should be a draw method that uses one generic texture for that

thobbsinteractive commented:
Also for some reason the Large textures are now being used when switched on

turican0 commented:
Don't worry, when I run the regression tests again, I'll run them on the rendered image as well, so I'll get to the error.
I'm more sorry that x64 still doesn't work, but maybe it's just a bug in the old renderer, hopefully it's just a minor thing.

thobbsinteractive commented:
You know, Tom I am pretty sure I can fix the cave roof render, I have seen that code before. It just needs to feed in a specific texture when it calls draw inverse triangle. There used to be a method to do that, I encountered it fixing the cave render in new render. Would you like me to fix it? or Would you rather I left it alone.

thobbsinteractive commented:
Fix the cave roof render it. It was only happening in the New Render so no regression test risk

thobbsinteractive commented:
I believe we can merge this now. If there are no more changes to this branch @turican0?

thobbsinteractive commented:
Cool. I will have to do this after work, so it will have to wait for a few
hours.

On Mon, 18 Jul 2022, 15:30 turican0, ***@***.***> wrote:

> Great, thanks for another correction. I think we can merge the branches.
> The most serious bug is that the game doesn't work in x64 now, but it
> didn't work in your branch before either. And I would only go into that in
> the next wave - along with regression tests.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/thobbsinteractive/magic-carpet-2-hd/pull/98#issuecomment-1187434105>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AETQQFDEWQQVLBHGVG3LNPLVUVL6LANCNFSM52ZUGNWQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

GrimSqueaker commented:
Nice. This is a great achievement, guys!

turican0 commented:
Plus x64 seems OK! It was just a minor incompatibility in the old Renderer, I added the fix to the new pull request. And another great news, regression testing only found the bug in level 22, but there are some random changes that make testing impossible. So there are no errors when running the levels. I might get around to regression testing the game in x64 tomorrow. Later I want to compare the renderers to see if their results are viable under all circumstances.
